subsys/usb/class/netusb/function_rndis.c

```diff
@@ -0,0 +1,1000 @@
+	u32_t tx_err;
+	u32_t rx_no_buf;
+
+	atomic_t notify_count;
+} rndis = {
+	.mac =  { 0x00, 0x11, 0x22, 0x33, 0x44, 0x66 },
+	.mtu = 1500, /* FIXME ? */
+	.media_status = RNDIS_OBJECT_ID_MEDIA_DISCONNECTED,
+	.state = UNINITIALIZED,
+	.skip_bytes = 0,
+};
+
+static u8_t manufacturer[] = { 't', 'e', 's', 't' };
+static u32_t drv_version = 1;
+
+static u32_t object_id_supported[] = {
+	RNDIS_OBJECT_ID_GEN_SUPP_LIST,
+	RNDIS_OBJECT_ID_GEN_HW_STATUS,
+	RNDIS_OBJECT_ID_GEN_SUPP_MEDIA,
+	RNDIS_OBJECT_ID_GEN_IN_USE_MEDIA,
+
+	RNDIS_OBJECT_ID_GEN_MAX_FRAME_SIZE,
+	RNDIS_OBJECT_ID_GEN_LINK_SPEED,
+	RNDIS_OBJECT_ID_GEN_BLOCK_TX_SIZE,
+	RNDIS_OBJECT_ID_GEN_BLOCK_RX_SIZE,
+
+	RNDIS_OBJECT_ID_GEN_VENDOR_ID,
+	RNDIS_OBJECT_ID_GEN_VENDOR_DESC,
+	RNDIS_OBJECT_ID_GEN_VENDOR_DRV_VER,
+
+	RNDIS_OBJECT_ID_GEN_PKT_FILTER,
+	RNDIS_OBJECT_ID_GEN_MAX_TOTAL_SIZE,
+	RNDIS_OBJECT_ID_GEN_CONN_MEDIA_STATUS,
+	RNDIS_OBJECT_ID_GEN_PHYSICAL_MEDIUM,
+#if defined(USE_RNDIS_STATISTICS)
+	/* Using RNDIS statistics puts heavy load on
+	 * USB bus, disable it for now
+	 */
+	RNDIS_OBJECT_ID_GEN_TRANSMIT_OK,
+	RNDIS_OBJECT_ID_GEN_RECEIVE_OK,
+	RNDIS_OBJECT_ID_GEN_TRANSMIT_ERROR,
+	RNDIS_OBJECT_ID_GEN_RECEIVE_ERROR,
+	RNDIS_OBJECT_ID_GEN_RECEIVE_NO_BUF,
+#endif /* USE_RNDIS_STATISTICS */
+	RNDIS_OBJECT_ID_802_3_PERMANENT_ADDRESS,
+	RNDIS_OBJECT_ID_802_3_CURR_ADDRESS,
+	RNDIS_OBJECT_ID_802_3_MCAST_LIST,
+	RNDIS_OBJECT_ID_802_3_MAX_LIST_SIZE,
+	RNDIS_OBJECT_ID_802_3_MAC_OPTIONS,
+};
+
+static int parse_rndis_header(const u8_t *buffer, u32_t buf_len)
+{
+	struct rndis_payload_packet *hdr = (void *)buffer;
+	u32_t len;
+
+	if (buf_len < sizeof(*hdr)) {
+		SYS_LOG_ERR("Too small packet len %u", buf_len);
+		return -EINVAL;
+	}
+
+	if (hdr->type != sys_cpu_to_le32(RNDIS_DATA_PACKET)) {
+		SYS_LOG_ERR("Wrong data packet type 0x%x",
+			    sys_le32_to_cpu(hdr->type));
+		return -EINVAL;
+	}
+
+	len = sys_le32_to_cpu(hdr->len);
+
+	if (len < sys_le32_to_cpu(hdr->payload_offset) +
+	    sys_le32_to_cpu(hdr->payload_len) +
+	    /*
+	     * Calculate additional offset since payload_offset is calculated
+	     * from the start of itself ;)
+	     */
+	    (u8_t *)&hdr->payload_offset - (u8_t *)hdr) {
+		SYS_LOG_ERR("Incorrect RNDIS packet");
+		return -EINVAL;
+	}
+
+	SYS_LOG_DBG("Parsing packet: len %u payload offset %u payload len %u",
+		    len, sys_le32_to_cpu(hdr->payload_offset),
+		    sys_le32_to_cpu(hdr->payload_len));
+
+	return len;
+}
+
+void rndis_clean(void)
+{
+	SYS_LOG_DBG("");
+
+	if (rndis.in_pkt) {
+		net_pkt_unref(rndis.in_pkt);
+
+		rndis.in_pkt = NULL;
+		rndis.in_pkt_len = 0;
+	}
+
+	rndis.skip_bytes = 0;
+}
+
+static void rndis_bulk_out(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+{
+	u8_t buffer[CONFIG_RNDIS_BULK_EP_MPS];
+	u8_t hdr_offset = 0;
+	u32_t len;
+	int ret;
+
+	usb_read(ep, NULL, 0, &len);
+
+	SYS_LOG_DBG("EP 0x%x status %d len %u", ep, ep_status, len);
+
+	usb_read(ep, buffer, len, NULL);
+
+	/* We already keep frame keeping with len, warn here about
```

Comment:
this comment would need to be rephrased.

Reply:
Fixed